Add EKS cluster name detection

crobert-1 opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This enhancement detects the k8s cluster name in EKS. The solution uses EC2 instance tags to determine the cluster name, which means it will only work on EC2 (as noted in documentation updates).

**Link to tracking Issue:** <Issue number if applicable>
Resolves #26794

**Testing:** <Describe what testing was performed and which tests were added.>
Manual testing in an EKS cluster when running on EC2 instances.

**Documentation:** <Describe the documentation added.>
Updated README and relevant metadata

crobert-1 committed:
Add k8s cluster name detection in EKS environment

Use ec2 instance tags to determine the EKS k8s cluster
name.

crobert-1 committed:
Update readme with permissions requirement.

crobert-1 committed:
Add permissions requirement

crobert-1 commented on processor/resourcedetectionprocessor/internal/aws/eks/metadata.yaml:
```diff
@@ -10,4 +10,8 @@ resource_attributes:
+  k8s.cluster.name:
+    description: The EKS cluster name. This attribute is currently only available when running on EC2 instances, and requires permission to run the EC2:DescribeInstances action.
+    type: string
     enabled: true
```
This will be enabled by default, let me know if it should be disabled by default.

jinja2 replied:
Users might have other processor adding this attribute, and since there is no official cluster name concept in k8s itself, they could've set it to some variation of eks name. I think it would be safer to leave it disabled initially.

dashpole commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 }
 
+func (d *detector) getClusterName() (string, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
```
should this use the context.Context from Detect?  Same comment for context.TODOs below.

crobert-1 replied:
I'm not the most well versed here on why, but experimentally directly using `ctx` is causing `context deadline exceeded`. Is it preferable to use it directly and set a longer timeout than using `context.TODO()`?

mx-psi replied:
I would vote for setting a longer timeout directly (the timeout comes from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/96d53a20663a0f593565fb84abbad99db3ce88da/processor/resourcedetectionprocessor/config.go#L38-L39C23 )

crobert-1 replied:
Done. I set the default timeout to 11 seconds when there's an EKS detector, which is the cutoff between success and failure in my demo environment for getting the cluster name.

dashpole commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
+func (d *detector) getClusterName() (string, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		d.logger.Debug("Failed to load default config", zap.Error(err))
```
generally, we should log or return the error, not both.  I would expect whatever invokes Detect() to handle the error appropriately (e.g. by logging it).  Applies here and below.

crobert-1 replied:
That's fair, I was mostly "inheriting" the style of the failure to detect an EKS environment in `Detect()`. In testing though, the messages are duplicated so I agree with your thought. I'll update.

dashpole commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 
+	clusterName, err := d.getClusterName()
+	if err != nil {
+		return pcommon.NewResource(), "", err
```
Will this be breaking for users that don't have the correct permissions?  Would returning `d.rb.Emit(), conventionss.SchemaURL, err` be better, or would the processor still use an empty resource in that case?

crobert-1 replied:
Good call. It's better to return as much as possible, especially when the other values here aren't dependent on the cluster name (and we've already successfully determined it's an EKS cluster). I'll update.

dashpole commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
+		}
+	}
+
+	d.logger.Debug("Could not find cluster name within EC2 instance tags")
```
should this return an error, rather than setting the cluster name to empty?

crobert-1 replied:
I've decided to just return an empty string, and then create an error within `getClusterName` if it's empty.

dashpole reviewed (commented):
LGTM overall, but someone from AWS should take a look.

crobert-1 committed:
Update dead links

crobert-1 committed:
Update broken links

crobert-1 commented on processor/resourcedetectionprocessor/README.md:
```diff
@@ -133,7 +133,7 @@ We map these environment variables to resource attributes as follows:
 | `HEROKU_SLUG_COMMIT`               | `heroku.release.commit`             |
 
-For more information, see the [Heroku cloud provider documentation](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud_provider/heroku.md) under the [OpenTelemetry specification semantic conventions](https://github.com/open-telemetry/opentelemetry-specification).
+For more information, see the [Heroku cloud provider documentation](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud_provider/heroku.md) under the [OpenTelemetry specification semantic conventions](https://github.com/open-telemetry/semantic-conventions).
```
```suggestion
For more information, see the [Heroku cloud provider documentation](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud-provider/heroku.md) under the [OpenTelemetry specification semantic conventions](https://github.com/open-telemetry/semantic-conventions).
```

crobert-1 committed:
Update processor/resourcedetectionprocessor/README.md

crobert-1 committed:
Changes requested by dashpole:

- Don't log and error, just error
- Return error for empty cluster name
- Return other information for other resources even if cluster name
detection fails.

crobert-1 committed:
Fix CICD failures

- Lint fix
- Test fix: Move accessory methods into utils helper for testability.

jinja2 commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -28,10 +32,15 @@ const (
 	authConfigmapNS             = "kube-system"
 	authConfigmapName           = "aws-auth"
+
+	clusterNameAwsEksTag = "aws:eks:cluster-name"
```
I would recommend to also check for the cluster name as part of the tag key `kubernetes.io/cluster/CLUSTER_NAME`. I think this tag is most likely to be available on older clusters with self-managed ASGs.

crobert-1 replied:
Thanks, I'll add.

jinja2 commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -114,3 +129,44 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
 }
+
+func (e eksDetectorUtils) getClusterName() (string, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		return "", err
+	}
+
+	ec2IMS := imds.NewFromConfig(cfg)
+	regionOutput, err := ec2IMS.GetRegion(context.TODO(), &imds.GetRegionInput{})
+	if err != nil {
+		return "", err
+	}
+	cfg.Region = regionOutput.Region
```
```suggestion
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithEC2IMDSRegion())
	if err != nil {
		return "", err
	}
```

Looking at the [default loader in the sdk](https://github.com/aws/aws-sdk-go-v2/blob/v1.21.0/config/config.go#L51), passing in this helper fn should make it query the region from imds.

crobert-1 replied:
Tested and it works, I'll update. Thanks!

jinja2 commented on processor/resourcedetectionprocessor/internal/aws/eks/detector.go:
```diff
@@ -114,3 +129,44 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
+	cfg.Region = regionOutput.Region
+
+	client := ec2.NewFromConfig(cfg)
+	describeInstances, err := client.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
```
I think this will query all ec2 instances, and we should get the tags for the local instance only. We can get the instance-id from imds and pass the instance id to `DescribeInstancesInput`.

crobert-1 replied:
Done.

crobert-1 committed:
Changes requested by jinja2

- Cleanup config initialization
- Use another possible tag for cluster name
- Only get local EC2 instance tags

crobert-1 committed:
Use proper context, set higher default timeout

crobert-1 committed:
Remove slices import since it's not introduced until later go versions

jinja2 commented on processor/resourcedetectionprocessor/README.md:
```diff
@@ -318,14 +318,18 @@ processors:
     * cloud.platform ("aws_eks")
+    * k8s.cluster.name
+
+Note: The kubernetes cluster name is only available when running on EC2 instances, and requires permission to run the `EC2:DescribeInstances` [action](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html).
```
We can also check if tags are available on the IMDS endpoint first. So if they are exposed on imds, we don't have to make the describe instance call. EC2 instances have a [setting to expose tags through IMDS](https://aws.amazon.com/about-aws/whats-new/2022/01/instance-tags-amazon-ec2-instance-metadata-service/).

crobert-1 replied:
I know it's possible, but it seemed more complicated to have two routes here. In my test environment `DescribeInstances` was allows, but `DescribeTags` wasn't by default, so this solution was what worked for me. 

Your suggestion would help in the case where permission is granted to run `DescribeTags` but not `DescribeInstances`. Do you think there's some level of likelihood that this would happen? If it is I'm happy to add the check, but I'd prefer to keep it simple if it's not something we expect to run into.

jinja2 replied:
Right, but this is trying to access tags from the local instance metadata service which shouldn't require an IAM role, just access to the local imds endpoint (the same way we fetched the instance id). But nevermind, the tags are not available on the imds endpoint by default and requires an additional config to be set in the ec2 instances.

crobert-1 committed:
Disable k8s cluster name by default

crobert-1 committed:
Run make generate

crobert-1 committed:
Fix test for disabled value

crobert-1 committed:
TESTING: Check if longer timeout will allow lint to succeed

crobert-1 committed:
TEST: Increase timeout to possibly allow linting pass

crobert-1 committed:
TEST: Increase timeout even more

crobert-1 committed:
TEST: More timeout increases

crobert-1 committed:
Keep increasing timeouts

crobert-1 committed:
TEST: Skip failing lint test

crobert-1 committed:
TEST: Skip entire cmd dir for linting

crobert-1 committed:
TEST: Increase memory garbage collection frequency

crobert-1 committed:
TEST: Increase memory allowance

crobert-1 committed:
Test: reduce memory usage

crobert-1 commented:
_Sorry for all of the updates here, linting is failing with OOM on the GitHub runners. I'm trying to find the best balance here of getting lint to complete successfully in the least amount of time possible._

crobert-1 committed:
TEST: GC set to 25

crobert-1 committed:
TEST: GC to 37

crobert-1 committed:
TEST: GC to 43

crobert-1 committed:
TEST: GC to 45

atoulme commented on .golangci.yml:
```diff
@@ -4,7 +4,7 @@ run:
 
   # timeout for analysis, e.g. 30s, 5m, default is 1m
-  timeout: 20m
+  timeout: 100m
```
is this still necessary?

crobert-1 replied:
My apologies, all of the changes you marked were a part of my testing to determine the failure point in the linting process. I'm going to revert as many of these changes as possible, and mark the review as `Ready for Review` when it's ready again.

atoulme replied:
No apologies necessary :) happy to review when it's good to go

atoulme commented on .golangci.yml:
```diff
@@ -21,6 +21,7 @@ run:
     - local
     - cmd/otelcontribcol
     - cmd/oteltestbedcol
+    - cmd
```
can be peeled off into a separate PR

atoulme commented on Makefile.Common:
```diff
@@ -23,8 +23,8 @@ SRC_ROOT := $(shell git rev-parse --show-toplevel)
-GOTEST_OPT?= -race -timeout 300s -parallel 4 --tags=$(GO_BUILD_TAGS)
-GOTEST_INTEGRATION_OPT?= -race -timeout 360s -parallel 4
+GOTEST_OPT?= -race -timeout 100m -parallel 4 --tags=$(GO_BUILD_TAGS)
+GOTEST_INTEGRATION_OPT?= -race -timeout 100m -parallel 4
```
same, please revert

atoulme commented on Makefile.Common:
```diff
@@ -174,7 +174,7 @@ fmt: $(GOIMPORTS)
 .PHONY: lint
 lint: $(LINT) checklicense misspell
-	$(LINT) run --allow-parallel-runners --verbose --build-tags integration --timeout=30m --path-prefix $(shell basename "$(CURDIR)")
+	GOGC=45 $(LINT) run --allow-parallel-runners --verbose --build-tags integration --timeout=100m --path-prefix $(shell basename "$(CURDIR)")
```
please revert

crobert-1 committed:
Revert timeout changes back to original values

crobert-1 committed:
Increase timeout

dmitryax commented:
Looks like your linter config change didn't work out. I see `make -C cmd/otelcontribcol lint` in the tests. The linter uses `cmd/otelcontribcol` as root directory

crobert-1 commented:
Superseded by #28649. `v1` of the `aws-sdk-go` library doesn't run into the linting issues, so I've decided to use that instead.
